app/src/test/java/com/cmput301f20t21/bookfriends/viewmodels/CreateAccountViewModelUnitTest.java

```diff
@@ -0,0 +1,56 @@
+package com.cmput301f20t21.bookfriends.viewmodels;
+
+import com.cmput301f20t21.bookfriends.entities.User;
+import com.cmput301f20t21.bookfriends.enums.SIGNUP_ERROR;
+import com.cmput301f20t21.bookfriends.fakes.callbacks.FakeFailCallbackWithMessage;
+import com.cmput301f20t21.bookfriends.fakes.callbacks.FakeSuccessCallback;
+import com.cmput301f20t21.bookfriends.fakes.repositories.FakeAuthRepository;
+import com.cmput301f20t21.bookfriends.fakes.repositories.FakeUserRepository;
+import com.cmput301f20t21.bookfriends.fakes.tasks.FakeSuccessTask;
+import com.cmput301f20t21.bookfriends.ui.login.CreateAccountViewModel;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class CreateAccountViewModelUnitTest {
+    private String uid = "uid";
+    private String username = "username";
+    private String email = "email";
+    private String password = "password";
+
+    @Mock
+    FakeAuthRepository mockAuthRepository;
+
+    @Mock
+    FakeUserRepository mockUserRepository;
+
+    @Mock
+    FakeSuccessCallback mockSuccessCallback;
+
+    @Mock
+    FakeFailCallbackWithMessage<SIGNUP_ERROR> mockFailCallbackWithMessage;
+
+    @Test
+    public void signUpSuccess() {
+        CreateAccountViewModel model = new CreateAccountViewModel(mockAuthRepository, mockUserRepository);
+        User user = new User(uid, username, email);
+
+        FakeSuccessTask<User> fakeUserTask = new FakeSuccessTask(user);
+
+
+        when(mockUserRepository.getByUsername(username)).thenReturn(fakeUserTask);
+
+
+        model.handleSignUp(username, email, password, mockSuccessCallback, mockFailCallbackWithMessage);
+        verify(mockFailCallbackWithMessage, times(1)).run(SIGNUP_ERROR.USERNAME_EXISTS);
```

Comment:
this shouldn't happen? If we sign up success then mockSuccessCallback has to be called not fail callback.

Reply:
I'm not sure about this, I followed the code in CreateAccountViewModel "userRepository.getByUsername(username) .addOnSuccessListener(user -> failCallback.run(SIGNUP_ERROR.USERNAME_EXISTS))" so I leave it as "mockFailCallbackWithMessage"

Reply:
<img width="742" alt="Screen Shot 2020-11-13 at 1 52 18 PM" src="https://user-images.githubusercontent.com/64346342/99120051-73faef00-25b7-11eb-971f-d07217346d45.png">

Reply:
Oh, I get why your confused here. `OnSuccessListener` doesn't mean sign up success. In this context, it means we successfully get a username from the database. In signup use case, this is a failure because the username already exist, so we call `failCallback` in success listener.
Only when the new username doesn't exist in the current database, then we can proceed registering him, therefore we actually handle sign up success in `OnFailureListener`.

Reply:
So i guess what this test is doing is mocking when a user exist and we sign up fail. Then it should be renamed to ``signupFail()`

Reply:
Got it !